### PR TITLE
chore: add eslint package to handle linting

### LINKED ISSUE #37254: [Task]: Add eslint browserify package

### PR DESCRIPTION
**Pull request by ayushpahwa**
## Description
We are shifting the linter engine for jshint to eslint. This PR adds the package needed for the same and also adds the required configs. The eslint feature is behing a rollout feature flag, it's currently at 0% right now.

 > [!NOTE]   > This PR is part of a series of [stacked diffs](https://newsletter.pragmaticengineer.com/p/stacked-diffs) and might have changes from it's parent PRs. Untill the blocking parent PRs are merged, this diff would show more changes than are relevant for this PR.  
Blocking PRs: - #36548 

Fixes #37254 

## Automation

/test js sanity

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11965021215>
> Commit: 579e4772f918936da292ba2815621ca80499c212
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11965021215&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.JS, @tag.Sanity`
> Spec:
> <hr>Fri, 22 Nov 2024 02:49:47 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Release Notes

- **New Features**
	- Integrated ESLint support alongside existing JSHint functionality for enhanced linting capabilities.
	- Added a new dependency for ESLint linting.

- **Improvements**
	- Enhanced error handling and reporting for linting errors, ensuring consistency across different linting tools.
	- Updated testing framework to accommodate multiple linter types without duplicating logic.
	- Introduced new functions for sanitizing and converting ESLint errors to the application's format.
	- Added support for structured cloning in the testing setup.
	- Improved logic for determining main actions and error handling in conditional expressions.

- **Bug Fixes**
	- Improved logic for handling lint errors, particularly in differentiating between ESLint and JSHint errors.

- **Documentation**
	- Updated comments for clarity on linting processes and configurations.

### CODE REVIEW

File: app/client/src/plugins/Linting/utils/getLintingErrors.ts [hunk 1]
```diff
@@ -161,6 +216,47 @@ const getLintErrorMessage = (
   }
 };
 
+function convertESLintErrorToAppsmithLintError(
+  eslintError: Linter.LintMessage,
+  script: string,
+  originalBinding: string,
+  scriptPos: Position,
+  //isJSObject = false,
```
Comment (KelvinOm) [inline app/client/src/plugins/Linting/utils/getLintingErrors.ts:224]: Сan we clean up this commented code and commented code below as well?

**Review by KelvinOm — DISMISSED** [review]
LGTM!

File: app/client/src/plugins/Linting/constants.ts [hunk 2]
```diff
@@ -1,38 +1,85 @@
...
+        "no-extend-native": "error",
+        "no-undef": "error",
+        "guard-for-in": "off",
+        "no-empty": "off",
+        strict: "off",
+        "no-unused-vars": [
+          "error",
+          { vars: "all", args: "all", ignoreRestSiblings: false },
+        ],
+        "no-cond-assign": "off",
```
[the remark this answers is not in the document]
Comment (ayushpahwa) [reply app/client/src/plugins/Linting/constants.ts:72]: I have checked these manually but it makes sense to add tests for these all, let me update.

File: app/client/src/plugins/Linting/utils/getLintingErrors.ts [hunk 3]
```diff
@@ -91,6 +92,60 @@ function generateLintingGlobalData(data: Record<string, unknown>) {
...
+  return lintErrors.reduce((result: Linter.LintMessage[], lintError) => {
+    // Ignored errors should not be reported
+    if (IGNORED_LINT_ERRORS.includes(lintError.ruleId || "")) return result;
+
+    /** Some error messages reference line numbers,
+     * Eg. Expected '{a}' to match '{b}' from line {c} and instead saw '{d}'
+     * these line numbers need to be re-calculated based on the binding location.
+     * Errors referencing line numbers outside the user's script should also be ignored
+     * */
+    let message = lintError.message;
```
Comment (sneha122) [inline app/client/src/plugins/Linting/utils/getLintingErrors.ts:108]: @ayushpahwa sanitizeESLintErrors function looks very similar to sanitizeJSHintErrors, do you think we can keep a single function but pass in linterType instead?

File: app/client/src/plugins/Linting/constants.ts [hunk 4]
```diff
@@ -1,38 +1,85 @@
 import { ECMA_VERSION } from "@shared/ast";
 import type { LintOptions } from "jshint";
 import isEntityFunction from "./utils/isEntityFunction";
+import type { Linter } from "eslint-linter-browserify";
```
Comment (sneha122) [inline app/client/src/plugins/Linting/constants.ts:4]: @ayushpahwa Where have we defined the flag that enables ESLint?

File: app/client/src/plugins/Linting/utils/getLintingErrors.ts (hunk 3 above)
Comment (ayushpahwa) [reply app/client/src/plugins/Linting/utils/getLintingErrors.ts:108]: The idea is to remove the jshint functions completely. If i add a conditional, this just makes the removal process complicated.

File: app/client/src/plugins/Linting/constants.ts (hunk 4 above)
Comment (ayushpahwa) [reply app/client/src/plugins/Linting/constants.ts:4]: he feature flag was added in a previous PR. Let me share all the PRs with you
PR1: https://github.com/appsmithorg/appsmith/pull/36543
PR2: https://github.com/appsmithorg/appsmith/pull/36548

Now we are on PR3, the current PR to add the eslint package

Next planned PRs
4) Add custom rules for floating promises
5) Remove jshint and related feature flags

**Pull request merged**